# 2. Variables and Statements

## 2.1. Variables

In [1]:
n = 17

In [2]:
pi = 3.141592653589793

In [3]:
message = 'And now for something completely different'

In [4]:
message

'And now for something completely different'

In [5]:
n + 25

42

In [6]:
2 * pi

6.283185307179586

In [7]:
round(pi)

3

In [8]:
len(message)

42

## 2.2 State diagrams

- n -> 17
- pi -> 3.141592653589793
- message -> 'And now for something completely different'

## 2.3. Variable names

In [9]:
million! = 1000000

SyntaxError: invalid syntax (347180775.py, line 1)

In [10]:
76trombones = 'big parade'

SyntaxError: invalid decimal literal (3381618747.py, line 1)

In [11]:
class = 'Self-Defence Against Fresh Fruit'

SyntaxError: invalid syntax (2069597409.py, line 1)

## 2.4. The import statement

In [12]:
import math

In [13]:
math.pi

3.141592653589793

In [14]:
math.sqrt(25)

5.0

In [15]:
math.pow(5, 2)

25.0

## 2.5. Expressions and statements

In [16]:
19 + n + round(math.pi) * 2

42

In [17]:
n = 17

In [18]:
import math

## 2.6. The print function

In [19]:
n + 1

18

In [20]:
n + 2
n + 3

20

In [21]:
print(n + 2)
print(n + 3)

19
20


In [22]:
print('The value of pi is approximately')
print(math.pi)

The value of pi is approximately
3.141592653589793


In [23]:
print('The value of pi is approximately', math.pi)

The value of pi is approximately 3.141592653589793


## 2.7. Arguments

In [24]:
int('101')

101

In [25]:
math.pow(5, 2)

25.0

In [26]:
int('101', 2)

5

In [27]:
round(math.pi, 3)

3.142

In [28]:
print('Any', 'number', 'of', 'arguments')

Any number of arguments


In [29]:
float('123.0', 2)

TypeError: float expected at most 1 argument, got 2

In [30]:
math.pow(2)

TypeError: pow expected 2 arguments, got 1

In [31]:
math.sqrt('123')

TypeError: must be real number, not str

## 2.8. Comments

In [32]:
# number of seconds in 42:42
seconds = 42 * 60 + 42

In [33]:
miles = 10 / 1.61   # 10 kilometers in miles

In [34]:
v = 8   # assign 8 to v

In [35]:
v = 8   # velocity in miles per hour

## 2.9. Debugging

In [36]:
million! = 1000000

SyntaxError: invalid syntax (347180775.py, line 1)

In [37]:
'126' / 3

TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [38]:
1 + 3 / 2

2.5

## 2.10. Glossary

- variable
- assignment statement
- state diagram
- keyword
- import statement
- module
- dot operator
- evaluate
- statement
- execute
- argument
- comment
- runtime error
- exception
- semantic error

## 2.11. Exercises

### 2.11.1. Ask a virtual assistant

- 파이썬에서 키워드는 왜 변수 이름으로 쓸 수 없어?
    > 좋은 질문이에요! 파이썬에서 \*\*키워드(keyword)\*\*를 변수 이름으로 쓸 수 없는 이유는, **키워드가 이미 파이썬 문법의 일부로 특별한 의미를 가지고 있기 때문**입니다.
    > 
    > ### 🔑 키워드란?
    > 
    > 키워드는 **파이썬이 문장을 이해하는 데 사용하는 예약어**입니다. 예를 들어, 다음과 같은 것들이 있죠:
    > * `if`, `else`, `while`, `for`, `def`, `class`, `return`, `import`, `True`, `False`, `None` 등등
    > 이런 키워드는 **파이썬이 특정한 문법 구조를 만들기 위해 쓰는 단어**들이에요. 그래서 이걸 변수 이름으로 쓰면, **파이썬이 헷갈려요**.
    > 
    > ### ❌ 예시로 살펴보기
    > 
    > ```python
    > if = 5
    > ```
    > 
    > 이 코드를 실행하면 에러가 납니다:
    > 
    > ```
    > SyntaxError: invalid syntax
    > ```
    > 
    > 왜냐하면 `if`는 파이썬에서 **조건문을 시작하는 데 쓰이는 문법 요소**이기 때문이에요. 그런데 우리가 이걸 변수 이름으로 쓰면 파이썬은:
    > > "어? `if` 다음에 조건이 나와야 하는데... 숫자 5가 왜 있지?"
    > 라고 생각하게 되는 거죠. 즉, **문법이 깨지는 거예요.**
    > 
    > ### 📌 정리하자면
    > 
    > * 키워드는 **파이썬 문법을 만드는 핵심 단어**라서
    > * 변수 이름으로 쓰면 **문법이 꼬일 수 있어서 금지**되어 있고
    > * 따라서 파이썬은 **이 키워드들을 "예약어"로 따로 막아놨습니다**
    > 
    > ### 🔍 팁: 키워드 목록 확인하는 방법
    > 
    > 파이썬에서 어떤 단어들이 키워드인지 알고 싶다면, `keyword` 모듈을 써보세요:
    > 
    > ```python
    > import keyword
    > print(keyword.kwlist)
    > ```
    > 
    > 이렇게 하면 현재 파이썬 버전에서 사용 중인 **모든 키워드 리스트**가 나옵니다.

- 왜 변수 이름으로 int, float, str을 사용하는 게 나쁜 방식이야?
    > 좋은 질문이에요! `int`, `float`, `str` 같은 이름은 **변수 이름으로 쓸 수는 있지만, 쓰면 안 좋은 이유**가 있어요. 그 이유는 이 이름들이 **파이썬의 내장 자료형 이름이기 때문**이에요.
    > 
    > ### 📌 `int`, `float`, `str`는 뭐지?
    > 
    > 이들은 파이썬의 **기본 자료형을 만드는 함수 또는 클래스**예요:
    > 
    > * `int()` → 정수로 변환
    > * `float()` → 실수로 변환
    > * `str()` → 문자열로 변환
    > 
    > 예시:
    > ```python
    > a = int("123")     # 문자열 "123"을 정수 123으로 바꿈
    > b = float("3.14")  # 문자열 "3.14"를 실수 3.14로 바꿈
    > c = str(456)       # 정수 456을 문자열 "456"으로 바꿈
    > ```
    > 
    > ### ❌ 그런데 이런 식으로 변수 이름을 쓰면?
    > 
    > ```python
    > int = 10
    > print(int("123"))
    > ```
    > 
    > 이걸 실행하면 어떻게 될까요?
    > 
    > ```
    > TypeError: 'int' object is not callable
    > ```
    > 
    > 왜냐하면, 우리가 `int`라는 이름을 **숫자 10을 담는 변수**로 덮어써버렸기 때문이에요. 이제 `int`는 더 이상 정수로 바꾸는 함수가 아니라 그냥 숫자 10이 되어버려서, `int("123")` 같은 코드를 쓸 수 없게 된 거예요.
    > 
    > ### 🔍 요약하자면
    > 
    > | 항목                    | 설명                                                 |
    > | --------------------- | -------------------------------------------------- |
    > | `int`, `float`, `str` | 파이썬 내장 함수/클래스 (자료형 변환에 사용)                         |
    > | 변수 이름으로 사용 가능?        | **가능하지만 추천하지 않음**                                  |
    > | 왜 나쁠까?                | 내장 기능을 **덮어쓰게 되어서** 다른 코드에서 에러 발생 가능               |
    > | 해결 방법                 | 변수 이름은 의미 있는 다른 이름으로 지정 (`num`, `value`, `text` 등) |
    > 
    > ### ✅ 좋은 변수 이름 예시
    > 
    > ```python
    > number = 10
    > price = 3.14
    > name = "Alice"
    > ```
    > 
    > ### 💡 참고: 내장 이름 확인하는 법
    > 
    > ```python
    > dir(__builtins__)
    > ```
    > 
    > 위 코드를 실행하면, 파이썬의 모든 내장 함수, 클래스 이름을 볼 수 있어요. 이 이름들은 **가급적 변수 이름으로 피해야 해요.**

- 파이썬에서 내장 함수는 어떤 게 있어?
    > 파이썬에는 \*\*내장 함수(built-in function)\*\*가 꽤 많아요. 이 함수들은 **어디서든 바로 사용할 수 있도록 기본으로 제공**되는 함수들이고, `import` 없이도 바로 쓸 수 있어요.
    > 
    > ## ✅ 대표적인 파이썬 내장 함수들
    > 
    > 다양한 목적에 따라 나눠서 정리해볼게요.
    > 
    > ### 📊 1. **기본 자료형 변환**
    > 
    > | 함수        | 설명       |
    > | --------- | -------- |
    > | `int()`   | 정수로 변환   |
    > | `float()` | 실수로 변환   |
    > | `str()`   | 문자열로 변환  |
    > | `bool()`  | 불리언으로 변환 |
    > | `list()`  | 리스트로 변환  |
    > | `tuple()` | 튜플로 변환   |
    > | `dict()`  | 딕셔너리로 변환 |
    > | `set()`   | 집합으로 변환  |
    > 
    > ### 🔢 2. **수학 관련**
    > 
    > | 함수             | 설명             |
    > | -------------- | -------------- |
    > | `abs()`        | 절댓값            |
    > | `pow(x, y)`    | x의 y제곱         |
    > | `round()`      | 반올림            |
    > | `max()`        | 가장 큰 값         |
    > | `min()`        | 가장 작은 값        |
    > | `sum()`        | 합계             |
    > | `divmod(a, b)` | (몫, 나머지) 튜플 반환 |
    > 
    > ### 🔁 3. **반복/순회 관련**
    > 
    > | 함수                       | 설명              |
    > | ------------------------ | --------------- |
    > | `range()`                | 숫자 범위 생성        |
    > | `enumerate()`            | (인덱스, 값) 튜플로 반환 |
    > | `zip()`                  | 여러 시퀀스를 묶음      |
    > | `map(func, iterable)`    | 각 요소에 함수 적용     |
    > | `filter(func, iterable)` | 조건을 만족하는 요소만 추출 |
    > | `sorted()`               | 정렬된 리스트 반환      |
    > | `reversed()`             | 반대 순서로 순회       |
    > 
    > ### 🧪 4. **자료형 검사 및 객체 정보**
    > 
    > | 함수                      | 설명            |
    > | ----------------------- | ------------- |
    > | `type()`                | 자료형 확인        |
    > | `isinstance(obj, type)` | 특정 타입인지 확인    |
    > | `id()`                  | 객체의 메모리 주소    |
    > | `len()`                 | 길이 반환         |
    > | `dir()`                 | 객체의 속성 리스트 확인 |
    > | `help()`                | 내장 도움말 출력     |
    > | `callable()`            | 호출 가능한지 확인    |
    > 
    > ### 📦 5. **입출력**
    > 
    > | 함수         | 설명        |
    > | ---------- | --------- |
    > | `print()`  | 출력        |
    > | `input()`  | 사용자 입력 받기 |
    > | `format()` | 문자열 포맷    |
    > 
    > ### 🧩 6. **기타**
    > 
    > | 함수                          | 설명                       |
    > | --------------------------- | ------------------------ |
    > | `eval()`                    | 문자열을 파이썬 코드로 실행          |
    > | `exec()`                    | 코드 블록 실행                 |
    > | `open()`                    | 파일 열기                    |
    > | `globals()` / `locals()`    | 전역/지역 네임스페이스 반환          |
    > | `any()`                     | 하나라도 참이면 True            |
    > | `all()`                     | 모두 참이면 True              |
    > | `bin()` / `oct()` / `hex()` | 2진수 / 8진수 / 16진수 문자열로 변환 |
    > | `chr()` / `ord()`           | 유니코드 ↔ 문자 변환             |
    > 
    > ## 🔍 전체 목록 확인 방법
    > 
    > 파이썬에서 내장 함수 전체 목록은 아래처럼 확인할 수 있어요:
    > 
    > ```python
    > import builtins
    > print(dir(builtins))
    > ```

- math 모듈에는 어떤 변수와 함수들이 있어?
    > `math` 모듈은 파이썬에서 수학적 연산을 위한 **유용한 함수들**을 제공하는 모듈이에요. 이 모듈을 사용하면 복잡한 수학적 계산을 쉽게 처리할 수 있죠. `math` 모듈에 포함된 **변수**와 **함수**들에 대해 정리해볼게요.
    > 
    > ### 🔢 **`math` 모듈 함수**
    > 
    > #### 1. **기본 수학 연산**
    > 
    > | 함수                  | 설명                            |
    > | ------------------- | ----------------------------- |
    > | `math.ceil(x)`      | `x`보다 크거나 같은 가장 작은 정수 반환 (올림) |
    > | `math.floor(x)`     | `x`보다 작거나 같은 가장 큰 정수 반환 (내림)  |
    > | `math.trunc(x)`     | `x`의 소수점 이하를 버림               |
    > | `math.fabs(x)`      | `x`의 절댓값                      |
    > | `math.factorial(x)` | `x`의 팩토리얼 (정수만)               |
    > | `math.gcd(x, y)`    | `x`와 `y`의 최대 공약수 (GCD)        |
    > | `math.lcm(x, y)`    | `x`와 `y`의 최소 공배수 (LCM)        |
    > 
    > #### 2. **수학 상수 및 파이 관련**
    > 
    > | 함수         | 설명                     |
    > | ---------- | ---------------------- |
    > | `math.pi`  | 원주율 (π)                |
    > | `math.e`   | 자연상수 (e)               |
    > | `math.tau` | 원의 한 바퀴 (τ = 2π)       |
    > | `math.inf` | 양의 무한대                 |
    > | `math.nan` | "Not a Number" (NaN) 값 |
    > 
    > #### 3. **지수 및 로그**
    > 
    > | 함수                  | 설명                                    |
    > | ------------------- | ------------------------------------- |
    > | `math.exp(x)`       | `e`의 `x`승                             |
    > | `math.log(x, base)` | `base`를 밑으로 하는 `x`의 로그 (밑을 생략하면 자연로그) |
    > | `math.log10(x)`     | 10을 밑으로 하는 `x`의 로그                    |
    > | `math.log2(x)`      | 2를 밑으로 하는 `x`의 로그                     |
    > | `math.pow(x, y)`    | `x`의 `y`승 (지수 연산)                     |
    > 
    > #### 4. **삼각 함수**
    > 
    > | 함수                 | 설명                     |
    > | ------------------ | ---------------------- |
    > | `math.sin(x)`      | `x`의 사인 (라디안 단위)       |
    > | `math.cos(x)`      | `x`의 코사인 (라디안 단위)      |
    > | `math.tan(x)`      | `x`의 탄젠트 (라디안 단위)      |
    > | `math.asin(x)`     | `x`의 아크사인 (결과는 라디안)    |
    > | `math.acos(x)`     | `x`의 아크코사인 (결과는 라디안)   |
    > | `math.atan(x)`     | `x`의 아크탄젠트 (결과는 라디안)   |
    > | `math.atan2(y, x)` | `y/x`의 아크탄젠트 (결과는 라디안) |
    > 
    > #### 5. **각도 변환**
    > 
    > | 함수                | 설명                 |
    > | ----------------- | ------------------ |
    > | `math.degrees(x)` | `x`를 라디안에서 각도로 변환  |
    > | `math.radians(x)` | `x`를 각도에서 라디안으로 변환 |
    > 
    > #### 6. **하이퍼볼릭 함수**
    > 
    > | 함수              | 설명             |
    > | --------------- | -------------- |
    > | `math.sinh(x)`  | `x`의 쌍곡선 사인    |
    > | `math.cosh(x)`  | `x`의 쌍곡선 코사인   |
    > | `math.tanh(x)`  | `x`의 쌍곡선 탄젠트   |
    > | `math.asinh(x)` | `x`의 아크쌍곡선 사인  |
    > | `math.acosh(x)` | `x`의 아크쌍곡선 코사인 |
    > | `math.atanh(x)` | `x`의 아크쌍곡선 탄젠트 |
    > 
    > #### 7. **기타 함수**
    > 
    > | 함수                    | 설명                         |
    > | --------------------- | -------------------------- |
    > | `math.modf(x)`        | `x`의 소수점 부분과 정수 부분을 튜플로 반환 |
    > | `math.isnan(x)`       | `x`가 NaN인지 확인              |
    > | `math.isfinite(x)`    | `x`가 유한한 값인지 확인            |
    > | `math.isinf(x)`       | `x`가 무한대인지 확인              |
    > | `math.copysign(x, y)` | `x`에 `y`의 부호를 복사           |
    > 
    > ### 🌍 **`math` 모듈 상수**
    > 
    > `math` 모듈에는 여러 가지 유용한 **수학 상수**들이 있습니다.
    > 
    > * `math.pi`: 원주율 $\pi$, 약 3.14159
    > * `math.e`: 자연 상수 $e$, 약 2.71828
    > * `math.tau`: $2 \pi$, 약 6.28318 (원 둘레와 관련된 값)
    > * `math.inf`: 양의 무한대
    > * `math.nan`: "Not a Number" (NaN, 정의되지 않은 값)
    > 
    > ### 📌 **사용 예시**
    > 
    > ```python
    > import math
    > 
    > # 1. 원주율 사용
    > print("pi:", math.pi)
    > 
    > # 2. 로그 계산
    > print("log10(1000):", math.log10(1000))
    > 
    > # 3. 제곱근 계산
    > print("sqrt(16):", math.sqrt(16))
    > 
    > # 4. 삼각 함수
    > angle_in_radians = math.radians(45)  # 45도를 라디안으로 변환
    > print("sin(45°):", math.sin(angle_in_radians))
    > 
    > # 5. 올림/내림
    > print("ceil(4.2):", math.ceil(4.2))  # 올림
    > print("floor(4.7):", math.floor(4.7))  # 내림
    > ```
    > 
    > ### ❓ **자주 쓰이는 함수들**
    > 
    > * `math.sqrt(x)`: 제곱근을 구할 때 자주 사용
    > * `math.sin()`, `math.cos()`, `math.tan()`: 삼각 함수는 기하학적 계산에 유용
    > * `math.exp()`, `math.log()`: 지수 및 로그 계산
    > * `math.gcd()`: 두 수의 최대공약수를 구할 때 유용
    > 
    > 이 외에도 `math` 모듈에는 다양한 함수들이 있어, **수학적 계산을 할 때** 매우 유용합니다.

- math 모듈 외에 파이썬 핵심 모듈이 뭐가 있어?
    > 파이썬에는 **math** 모듈 외에도 매우 유용하고 중요한 **핵심 모듈**들이 많아요. 이런 모듈들은 파이썬의 **표준 라이브러리**에 포함되어 있어서, 추가 설치 없이 바로 사용할 수 있습니다.
    > 
    > 다음은 **주요 핵심 모듈**들을 소개할게요:
    > 
    > ## 1. **`os` 모듈** (운영체제 관련)
    > 
    > `os` 모듈은 운영체제와 관련된 다양한 기능을 제공합니다. 파일과 디렉토리 작업, 환경 변수 등을 처리할 수 있어요.
    > 
    > ### 주요 기능:
    > 
    > * `os.getcwd()` → 현재 작업 디렉토리 반환
    > * `os.chdir(path)` → 작업 디렉토리 변경
    > * `os.listdir(path)` → 디렉토리 내 파일 목록
    > * `os.rename(old_name, new_name)` → 파일/디렉토리 이름 변경
    > * `os.remove(path)` → 파일 삭제
    > * `os.mkdir(path)` → 디렉토리 생성
    > * `os.environ` → 환경 변수 다루기
    > 
    > ## 2. **`sys` 모듈** (시스템 관련)
    > 
    > `sys` 모듈은 파이썬 인터프리터와 관련된 작업을 다룹니다. 프로그램 실행과 관련된 다양한 정보를 얻거나, 파이썬 인터프리터에 영향을 줄 수 있어요.
    > 
    > ### 주요 기능:
    > 
    > * `sys.argv` → 명령줄 인자 목록
    > * `sys.exit()` → 프로그램 종료
    > * `sys.version` → 파이썬 버전 확인
    > * `sys.path` → 모듈 검색 경로
    > 
    > ## 3. **`datetime` 모듈** (날짜와 시간 처리)
    > 
    > `datetime` 모듈은 날짜와 시간을 처리하는 데 매우 유용합니다. 날짜 차이를 계산하거나, 포맷을 지정하여 출력할 수 있어요.
    > 
    > ### 주요 기능:
    > 
    > * `datetime.datetime.now()` → 현재 날짜와 시간
    > * `datetime.datetime.strptime(date_string, format)` → 문자열을 `datetime` 객체로 변환
    > * `datetime.timedelta(days=0, seconds=0)` → 날짜 차이 계산
    > * `datetime.date.today()` → 현재 날짜
    > * `datetime.datetime.today()` → 현재 날짜와 시간
    > 
    > ## 4. **`random` 모듈** (난수 생성)
    > 
    > `random` 모듈은 무작위 숫자 생성 및 셔플, 샘플링 등을 할 수 있는 함수들을 제공합니다.
    > 
    > ### 주요 기능:
    > 
    > * `random.random()` → 0에서 1 사이의 랜덤 실수
    > * `random.randint(a, b)` → `a`와 `b` 사이의 랜덤 정수
    > * `random.choice(sequence)` → 시퀀스에서 랜덤한 항목 선택
    > * `random.shuffle(sequence)` → 시퀀스 항목의 순서 랜덤하게 섞기
    > * `random.sample(sequence, k)` → 시퀀스에서 중복 없이 `k`개 항목 샘플링
    > 
    > ## 5. **`json` 모듈** (JSON 처리)
    > 
    > `json` 모듈은 JSON 데이터의 직렬화(파이썬 객체를 JSON 문자열로 변환)와 역직렬화(JSON 문자열을 파이썬 객체로 변환)를 처리합니다.
    > 
    > ### 주요 기능:
    > 
    > * `json.dumps(obj)` → 파이썬 객체를 JSON 문자열로 변환
    > * `json.loads(json_string)` → JSON 문자열을 파이썬 객체로 변환
    > * `json.dump(obj, file)` → 파이썬 객체를 JSON 형식으로 파일에 기록
    > * `json.load(file)` → 파일에서 JSON 데이터를 읽어 파이썬 객체로 변환
    > 
    > ## 6. **`re` 모듈** (정규 표현식)
    > 
    > `re` 모듈은 문자열 패턴 매칭을 위해 정규 표현식을 다루는 기능을 제공합니다.
    > 
    > ### 주요 기능:
    > 
    > * `re.match(pattern, string)` → 문자열이 패턴에 맞는지 확인
    > * `re.search(pattern, string)` → 문자열에서 패턴 찾기
    > * `re.findall(pattern, string)` → 문자열에서 패턴과 일치하는 모든 부분 찾기
    > * `re.sub(pattern, replacement, string)` → 패턴을 다른 문자열로 치환
    > 
    > ## 7. **`collections` 모듈** (특수한 컬렉션 데이터 구조)
    > 
    > `collections` 모듈은 기본적인 `list`, `dict`, `tuple` 외에도 유용한 데이터 구조들을 제공합니다.
    > 
    > ### 주요 기능:
    > 
    > * `collections.Counter` → 항목의 개수를 셀 수 있는 딕셔너리 서브클래스
    > * `collections.defaultdict` → 기본값을 제공하는 딕셔너리
    > * `collections.namedtuple()` → 이름이 붙은 튜플 객체 생성
    > * `collections.deque` → 양쪽 끝에서 빠르게 append, pop이 가능한 큐
    > 
    > ## 8. **`itertools` 모듈** (반복자 관련 도구)
    > 
    > `itertools` 모듈은 반복자(iterator)와 관련된 여러 가지 유용한 함수를 제공합니다. 순열, 조합, 카르타시안 곱 등 다양한 반복 작업을 처리할 수 있어요.
    > 
    > ### 주요 기능:
    > 
    > * `itertools.count(start=0, step=1)` → 시작 값부터 무한 반복 생성
    > * `itertools.cycle(iterable)` → 순차적으로 반복
    > * `itertools.permutations(iterable, r)` → 순열 생성
    > * `itertools.combinations(iterable, r)` → 조합 생성
    > 
    > ## 9. **`subprocess` 모듈** (서브프로세스 관리)
    > 
    > `subprocess` 모듈은 외부 프로세스를 생성하고 관리하는 데 사용됩니다. 다른 프로그램을 호출하거나 파이썬 코드에서 외부 명령어를 실행할 수 있어요.
    > 
    > ### 주요 기능:
    > 
    > * `subprocess.run()` → 외부 명령어 실행
    > * `subprocess.Popen()` → 프로세스 실행 후, 제어권 반환
    > 
    > ## 10. **`socket` 모듈** (네트워크 관련)
    > 
    > `socket` 모듈은 네트워크 소켓을 통해 통신할 수 있게 해줍니다. 서버와 클라이언트 간의 연결을 설정하거나 데이터를 송수신할 수 있어요.
    > 
    > ### 주요 기능:
    > 
    > * `socket.socket()` → 소켓 객체 생성
    > * `socket.connect()` → 서버에 연결
    > * `socket.send()` → 데이터 송신
    > * `socket.recv()` → 데이터 수신
    > 
    > ### 📌 **기타 중요한 파이썬 모듈**
    > 
    > * **`socket`**: 네트워크 통신
    > * **`shutil`**: 파일 및 디렉토리 작업
    > * **`email`**: 이메일 처리
    > * **`sqlite3`**: SQLite 데이터베이스 작업
    > * **`csv`**: CSV 파일 읽기/쓰기
    > * **`hashlib`**: 해시 함수 (MD5, SHA 등)
    > * **`pdb`**: 파이썬 디버거
    > * **`logging`**: 로그 기록
    > * **`argparse`**: 명령줄 인자 처리
    > 
    > ### 🛠️ **모듈 사용 예시**
    > 
    > ```python
    > # os 모듈 예시: 현재 작업 디렉토리 확인
    > import os
    > print(os.getcwd())  # 현재 작업 디렉토리 출력
    > 
    > # sys 모듈 예시: 파이썬 버전 확인
    > import sys
    > print(sys.version)  # 파이썬 버전 출력
    > 
    > # json 모듈 예시: 파이썬 객체를 JSON 문자열로 변환
    > import json
    > data = {"name": "Alice", "age": 25}
    > json_str = json.dumps(data)
    > print(json_str)  # {"name": "Alice", "age": 25}
    > ```
    > 
    > ### **정리**
    > 
    > 파이썬에서 **핵심 모듈**들은 다양한 작업을 효율적으로 처리할 수 있게 도와주는 도구들이에요. 기본적인 **파일 처리**, **수학적 연산**, **네트워크**, **문자열 처리** 등 다양한 분야에서 사용됩니다. 이런 모듈을 잘 활용하면, 파이썬을 더욱 강력하게 사용할 수 있습니다.

### 2.11.2. Exercise

In [39]:
17 = n

SyntaxError: cannot assign to literal here. Maybe you meant '==' instead of '='? (1575534540.py, line 1)

In [40]:
x = y = 1

In [41]:
n = 17;

In [42]:
n = 17.

In [43]:
print(n).

SyntaxError: invalid syntax (2959997231.py, line 1)

In [44]:
import maath

ModuleNotFoundError: No module named 'maath'

### 2.11.3. Exercise

In [45]:
radius = 5  # centimeters
volume = (4 / 3) * math.pi * (radius ** 3)  # cubic centimeters
print(volume)

523.5987755982989


In [46]:
x = 42
print((math.cos(42) ** 2) + (math.sin(42) ** 2))

1.0


In [47]:
print(math.e ** 2)
print(math.pow(math.e, 2))
print(math.exp(2))

7.3890560989306495
7.3890560989306495
7.38905609893065
